### PR TITLE
Add `ApolloClient.retryOnError(Boolean)`

### PR DESCRIPTION
**Pull request by martinbonnin**
So that took a while but I'm quite happy with this solution. It's a lot more flexible that the previous WebSocket only and in the end it's just a few lines of code meaning someone wanting to add custom logic shouldn't need copy/pasting too many things.

### CODE REVIEW

File: tests/websockets/src/jvmTest/kotlin/incubating/RetryTest.kt [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import sample.server.TagSubscription
+import sample.server.ZeroQuery
+import kotlin.test.assertIs
+
+class RetryTest {
+  @Test
+  fun retryIsWorking() = runTest(skipDelays = false) {
+    // Hopefully it's really a free port, if not this test will fail
+    val freePort = 8392
+    var sampleServer = SampleServer(freePort, "tag1")
```
Comment (BoD) [inline tests/websockets/src/jvmTest/kotlin/incubating/RetryTest.kt:26]: We could extract the port from sampleServer.subscriptionsUrl(). But the hardcoded port is most probably fine.

File: libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/interceptor/RetryOnErrorInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      attempt++
+
+      delay(2.0.pow(attempt).seconds)
+      cause is RetryException
+    }
+  }
+}
+
+
+private fun ApolloException.isTerminalAndRecoverable(): Boolean {
```
Comment (BoD) [inline libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/interceptor/RetryOnErrorInterceptor.kt:61]: Would it make sense to have this logic as a configurable lambda on `ApolloResponse`? So one could check for `exception` but also `errors` or even `data`.

**Review by BoD — APPROVED** [review]
Very nice! 🎉

File: tests/websockets/src/jvmTest/kotlin/incubating/RetryTest.kt (hunk 1 above)
Comment (martinbonnin) [reply tests/websockets/src/jvmTest/kotlin/incubating/RetryTest.kt]: Excellent point, done in https://github.com/apollographql/apollo-kotlin/pull/5685/commits/297950d86f4e5b21421b323b9a9d4bce26f24e9b

File: libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/interceptor/RetryOnErrorInterceptor.kt (hunk 2 above)
Comment (martinbonnin) [reply libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/interceptor/RetryOnErrorInterceptor.kt:53]: I have bikeshedded the API a bit:

* `retryNetworkErrors` -> `retryOnError` because it's shorter and pronounces better IMO
* There is now a required `defaultRetryOnError` in `RetryOnErrorInterceptor` so as to not break the symmetry of `RetryOnErrorInterceptor`

Overall I'm not sure yet how much retry logic belongs to user land vs something that we can provide with a bunch of parameters (the lambda you're proposing, + exponential backoff + Android Connectivity manager, maybe auth, etc...). My hunch is that it's certainly something that belongs to user land but maybe after some usage we can identify a minimal set of params that would serve the most use cases.

tldr; I would make `setRetryOnErrorInterceptor` API stable soon but wait a lot more for `RetryOnErrorInterceptor`

Comment (martinbonnin) [reply libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/interceptor/RetryOnErrorInterceptor.kt:53]: I ended up changing thing again [here](https://github.com/apollographql/apollo-kotlin/pull/5690/files#diff-aa6dd8789c20afff5b1d558e73067359cd06b9313363cc9dc764a2c551611ac4R360). The `RetryOnErrorInterceptor` is now always installed by default. 

`ApolloClient.Builder` now takes a `retryOnError((ApolloRequest) -> Boolean)` to set a default `retryOnError` when not specified. By default it is false. You can see it in action  [here](https://github.com/apollographql/apollo-kotlin/pull/5690/files#diff-956eb6b81f65aa82149cac71ee464043a3c5b605f8bbe886417469241f3b48b9L28)

Comment (BoD) [reply libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/interceptor/RetryOnErrorInterceptor.kt:53]: so retrying subscriptions is no longer the default, it needs to be explicit, right?

Comment (martinbonnin) [reply libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/interceptor/RetryOnErrorInterceptor.kt:53]: It never has been. In that last [pr](https://github.com/apollographql/apollo-kotlin/pull/5690/files#diff-aa6dd8789c20afff5b1d558e73067359cd06b9313363cc9dc764a2c551611ac4R360), the `RetryOnErrorInterceptor` is always used and opt-in is made with `ApolloClient.Builder.retryOnError {}`while in previous attempts, opt-in was made with `ApolloClient.Builder.retryOnErrorInterceptor()`

Comment (BoD) [reply libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/interceptor/RetryOnErrorInterceptor.kt:53]: All right - SGTM 👍

**Review by BoD — APPROVED** [review]
